Display ahead and behind information in the push button

pmiossec opened the pull request:
when the current branch is tracking a remote one.

## Proposed changes

* Display the count of unpushed commits and not integrated commit in the '2↑ 5↓' format
* Update the tooltip of the "push" button to explain the numbers
* Change the color of the "push" button if commits need to be integrated (or force pushed)

This change was made to improve understanding by the user of what Git will do when it will push.
It's a special ask by a git/gitextensions newcomer that liked the Visual Studio way to present data (because he don't really used eanough git to understand the git graph displayed in GitExtensions)

Requires Git >=2.5 (there is a control).

## Screenshots

### Before
A normal push button:

![image](https://user-images.githubusercontent.com/460196/50453501-482ddc00-0941-11e9-8b88-71666a6f406b.png)

### After
When there is something to push:

![image](https://user-images.githubusercontent.com/460196/50453777-5aa91500-0943-11e9-8ae7-e367b0f8406f.png)


When the branch is behind (not up to date):

![image](https://user-images.githubusercontent.com/460196/50453703-d0f94780-0942-11e9-94d1-96d70e840ae9.png)


When the branch is ahead and also behind commits (that required an integration --rebase or merge-- or a force push):

![image](https://user-images.githubusercontent.com/460196/50453750-27ff1c80-0943-11e9-87f1-769d23ff7c42.png)

The setting to disable the text in the "push" button for thoose that don't like button that resize themself, like the commit one (but that don't disable the tooltip and the color change in the button):

![image](https://user-images.githubusercontent.com/460196/50453804-7f04f180-0943-11e9-85e0-5c8a51c86e5f.png)

PS: I reused the "unstage" icon but I would prefer a new color (orange?) but I'm not a gimp expert. I will have a look later... (Or if someone could help).

## Test methodology

- manual tests

## Test environment(s) <!-- Remove any that don't apply -->

- GIT 2.20
- Windows 10

pmiossec commented:
I also would like to display the same information for all the local branches in the left panel.
I mainly achieved it in https://github.com/pmiossec/gitextensions/tree/display_ahead_behind_left_panel
but even if the `Text` property of the nodes are set, the tree is not refreshed.

If someone know how to do it. I already played with `BeginEdit()` and `EndEdit()` without success :(

vbjay commented on GitUI/CommandsDialogs/FormBrowse.cs:
```diff
@@ -373,6 +375,75 @@ void ManageWorktreeSupport()
+                return;
+            }
+
+            var result = UICommands.GitModule.RunGitCmd(aheadBehindGitCommand + currentBranch);
```
Use gitargumentbuilder.

vbjay replied:
```suggestion
    var cmd= new GitArgumentBuilder("for-each-ref")
    {
     "--format=\"%(push:track) %(refname:short)\"",
     "refs/heads/",
     currentBranch
    };
            var result = UICommands.GitModule.RunGitCmd(cmd);
```

Something like that.

vbjay replied:
I updated the wiki.

https://github.com/gitextensions/gitextensions/wiki#coding-guide

vbjay commented on GitUI/CommandsDialogs/FormBrowse.cs:
```diff
@@ -373,6 +375,75 @@ void ManageWorktreeSupport()
+        private readonly TranslationString _aheadCommitsToPush = new TranslationString("{0} new commit(s) will be pushed");
+        private readonly TranslationString _behindCommitsTointegrateOrForcePush = new TranslationString("{0} commit(s) should be integrated (or will be lost if force pushed)");
+
+        private const string aheadBehindGitCommand = "for-each-ref --format=\"%(push:track) %(refname:short)\" refs/heads/";
```
Remove this line.

gerhardol commented on GitCommands/Git/GitVersion.cs:
```diff
@@ -108,6 +109,8 @@ int Get(IReadOnlyList<int> values, int index)
 
         public bool RaceConditionWhenGitStatusIsUpdatingIndex => this < v2_0_1;
 
+        public bool SupportAheadBehindData => this >= v2_5_0;
```
Officially 2.11 is required but I guess it is better to have the check...

gerhardol commented on GitUI/CommandsDialogs/FormBrowse.cs:
```diff
@@ -476,6 +547,8 @@ protected override void OnLoad(EventArgs e)
                     .FileAndForget();
             };
 
+            DisplayAheadBehindInformation();
```
Noting: Similar handling as in #5777

RussKie commented on GitUI/CommandsDialogs/FormBrowse.cs:
```diff
@@ -373,6 +375,75 @@ void ManageWorktreeSupport()
+                return;
+            }
+
+            toolStripButtonPush.DisplayStyle = ToolStripItemDisplayStyle.Image;
```
I am not content with adding more code to `FormBrowse`, it is already several thousand lines long.
This functionality seems fairly self contained and could be abstracted in a form of a usercontrol/custom toolstrip button.

pmiossec commented:
A beginning...
* Add some unit tests (will surely add more)
* create a helper to encapsulate ahead/behind logic
* create a control to encapsulate ahead/behind logic for the push button

RussKie commented:
Looking forward to it, thank you

On Fri, 28 Dec 2018 at 20:42, Philippe Miossec <notifications@github.com>
wrote:

> A beginning...
>
>    - Add some unit tests (will surely add more)
>    - create a helper to encapsulate ahead/behind logic
>    - create a control to encapsulate ahead/behind logic for the push
>    button
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/5994#issuecomment-450328835>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEMyXk5hvNtG2qocGD1d3g_7T47Gso2dks5u9eeigaJpZM4Zh-86>
> .
>

RussKie commented on GitUI/CommandsDialogs/ToolStripPushButton.cs:
```diff
@@ -0,0 +1,77 @@
+
+        public void DisplayAheadBehindInformation()
+        {
+            if (!GitVersion.Current.SupportAheadBehindData)
```
This makes the class untestable, let's inject the flag via `Initialize` method like so:
```
        public void Initialize(IGitModule gitModule, bool supportsAheadBehindData)
        {
            ...
        }
```

RussKie commented on GitUI/CommandsDialogs/AheadBehindHelper.cs:
```diff
@@ -0,0 +1,88 @@
+            new Regex(@"^\[(ahead (?<ahead>\d+))?(, )?(behind (?<behind>\d+))?\] (?<branch>.*)$",
+                RegexOptions.Compiled | RegexOptions.Multiline);
+
+        public AheadBehindHelper(IGitModule gitModule)
```
I think in this instance we can avoid passing `IGitModule` - it is really used to execute a git command, that in turn is facilitated by an instance of `IExecutable` on an instance of `GitModule`.

RussKie commented:
I pushed a set of refactors with my proposals for API. Keen to know what you think.
NB: I haven't run it.

pmiossec commented:
>I pushed a set of refactors with my proposals for API. Keen to know what you think.
>NB: I haven't run it.

I tried to do the last fixes to run it.
It run well.
But fixing the tests is a real pain. Mocking `IExecutable` to get the output command result is a lot more difficult than before (due to an extension method that nsubstitute can't mock...). I didn't know how to do it for the moment and will try to have another look later...

RussKie commented:
I'll have a look, hopefully today

On Tue, Jan 1, 2019, 10:31 AM Philippe Miossec <notifications@github.com
wrote:

> I pushed a set of refactors with my proposals for API. Keen to know what
> you think.
> NB: I haven't run it.
>
> I tried to do the last fixes to run it.
> It run well.
> But fixing the tests is a real pain. Mocking IExecutable to get the
> output command result is a lot more difficult than before (due to an
> extension method that nsubstitute can't mock...). I didn't know how to do
> it for the moment and will try to have another look later...
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/5994#issuecomment-450698979>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEMyXvG7_0unA8zKqDJgKfqbJoCDfZh-ks5u-p5mgaJpZM4Zh-86>
> .
>

pmiossec commented:
>I'll have a look, hopefully today

Because `GetOutputAsync()` is an extension method (and so can't be mocked by nsubtitute), we have to mock everything is used in this method. It was easier to mock before because the return type was just the command output as a string.

I pushed my WIP for you.
Perhaps we could introduce another more high level interface that wrap `IExecutable` with a method that take the arguments and return the output string...

RussKie commented:
> Because `GetOutputAsync()` is an extension method (and so can't be mocked by nsubtitute), we have to mock everything is used in this method. It was easier to mock before because the return type was just the command output as a string.

This clearly demonstrates that the implemented abstraction is incorrect 😎 (I should have noticed that earlier).
We are trying to implement unit tests, but the current implementation demands integration tests due to the following piece of code:
```
        public void Initialize(Func<IExecutable> getGitExecutable, bool supportsAheadBehindData)
        {
            _aheadBehindDataProvider = new AheadBehindDataProvider(getGitExecutable);
        }
```

I have changed this and provided some sample tests, please have a look.

. . .

If it possible, do you think you could use "_undescore_separated_naming_convention_" for tests instead of "_PascalCaseBasedNamingConvention_" as it makes it slightly easier to read.

pmiossec commented:
>I have changed this and provided some sample tests, please have a look.

I had a very quick look but it seems that you solved the easiest part.

In the push button control tests, I reused the mocks done in the 'aheadbehindprovider' because it was easier. But the problem still need to be solved for the tests of the provider, this time with the 'IExecute' 😕

RussKie commented:
OK, let me get back to you

On Jan 3, 2019 11:07 AM, "Philippe Miossec" <notifications@github.com>
wrote:

I have changed this and provided some sample tests, please have a look.

I had a very quick look but it seems that you solved the easiest part.

In the push button control tests, I reused the mocks done in the
'aheadbehindprovider' because it was easier. But the problem still need to
be solved for the tests of the provider, this time with the 'IExecute' 😕

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/gitextensions/gitextensions/pull/5994#issuecomment-451023935>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/AEMyXgXIc2sSrOVwgOWPjNrEYaNkrrKSks5u_UnBgaJpZM4Zh-86>
.

RussKie commented:
Perhaps I'm missing something or we're talking about different things... If you could clarify what you're having troubles with, I'll try to help.

**Unit tests:**
* `ToolStripPushButton` only depends on `IAheadBehindDataProvider`, so tests for `ToolStripPushButton.Initialize()` would only involve mocking of `IAheadBehindDataProvider`. There should be no unit tests depending on `IExecutable`.
* `AheadBehindDataProvider` only depends on `IExecutable`, so it should be mocked as well.
In fact `IExecutable.GetOutput()` is the only method that needs mocking.

**Integration tests:**
These are quite different, and would involve building a test git repo, running real git commands and rendering a UI control. 
We can certainly do this, though at this stage I'm unsure these would provide a significant value.

pmiossec commented:
>AheadBehindDataProvider only depends on IExecutable, so it should be mocked as well.
In fact IExecutable.GetOutput() is the only method that needs mocking.

Unfortunately, no, you can't mock `GetOutput()` (or `GetOutputAsync()`) that are extension methods (at least with nsubstitute). The only method in IExcutable is `Start()` so you have in fact to mock everything is used in `GetOutputAsync()` which is not an easy thing to do. You have to mock `Iprocess` to return the good `OutputStream`, instanciate a MemoryStream,.... only to get the output string 😢

It's doable but I feel that...

> [...] the implemented abstraction is incorrect

😆

pmiossec commented:
>Unfortunately, no, you can't mock

When I say that, it is in fact the code I wrote, so that compile but there is a runtime exception when the test is run.

RussKie commented:
Ah sorry, I didn't look at the actual interface. Thank you for the
clarifications. I'll have a look tonight

On Thu, Jan 3, 2019, 7:54 PM Philippe Miossec <notifications@github.com
wrote:

> Unfortunately, no, you can't mock
>
> When I say that, it is in fact the code I wrote, so that compile but there
> is a runtime exception when the test is run.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/5994#issuecomment-451084425>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEMyXjTLowkmNu_H61lIMZOyF9ydv7pnks5u_cU1gaJpZM4Zh-86>
> .
>

RussKie commented:
@drewnoakes could you recall the reason for creating extension methods instead of implementing methods on the actual class?

RussKie commented:
@pmiossec I pushed an update, please have a look. 

With extension methods it is certainly more complicated and required few slight API tweaks, but overall it wasn't too bad.

RussKie commented on GitCommands/Git/AheadBehindDataProvider.cs:
```diff
@@ -0,0 +1,107 @@
+
+        public IDictionary<string, AheadBehindData> GetData(string branchName = "*")
+        {
+            return GetData(GitModule.SystemEncoding, branchName);
```
I realised after I pushed it in that I did it incorrectly

```suggestion
            return GetData(null, branchName);
```

RussKie commented on GitCommands/Git/ExecutableExtensions.cs:
```diff
@@ -16,7 +16,7 @@ namespace GitCommands
     {
         private static readonly Regex _ansiCodePattern = new Regex(@"\u001B[\u0040-\u005F].*?[\u0040-\u007E]", RegexOptions.Compiled);
-        private static readonly Encoding _defaultOutputEncoding = GitModule.SystemEncoding;
+        private static readonly Lazy<Encoding> _defaultOutputEncoding = new Lazy<Encoding>(() => GitModule.SystemEncoding, false);
```
If we don't make it lazy, it gets executed and that has a snowball effect - a hardcoded instance of `GitModule` gets created, that executes a git command to determine a default system encoding...

pmiossec commented:
@RussKie I have fixed the test and rebased th branch on `upstream` to fix the merge conflict...

Tell me if you want me to squash all the commits now...

* I have also changed the color of the push even when the branch is behind (even if no new commit to push). I think it makes more sense. Color changed == force pushed required or rebase/merge.

* There is also the commit adding the same ahead/behind data displayed in the left panel :

![image](https://user-images.githubusercontent.com/460196/50668686-e6184b00-0fc0-11e9-8844-89390e4d7ba6.png)

RussKie commented on UnitTests/GitCommandsTests/Git/AheadBehindDataProviderTests.cs:
```diff
@@ -0,0 +1,167 @@
+            data.Should().BeNull();
+        }
+
+        private void SetResultOfGitCommand(string result)
```
Please move it after all tests

RussKie commented on GitUI/BranchTreePanel/RepoObjectsTree.cs:
```diff
@@ -425,5 +427,38 @@ private void OnNodeDoubleClick(object sender, TreeNodeMouseClickEventArgs e)
+                return;
+            }
+
+            var aheadBehindData = new AheadBehindDataProvider(() => Module.GitExecutable).GetData();
```
`AheadBehindDataProvider` is stateless, so we don't need to re-instantiate it every time.
Please check if it can be instantiated once and then reused, maybe at the time `Module` is set.

RussKie commented:
Good job! I think it reads pretty well. 

The diff is clean, I suggest squashing it. Maybe separate into two commits - one for the button (with all infra) and another for the left panel changes.

I'm a bit concerned with how `AheadBehindDataProvider` is set in the left panel. 
Other than that I'm quite happy how it turned. Hope you're pleased too.

pmiossec commented:
done.

RussKie commented:
what about this?
https://github.com/gitextensions/gitextensions/pull/5994/files#r244976171

pmiossec commented:
this time, done ;)

RussKie commented:
@vbjay is it ok to merge?

gerhardol commented on GitCommands/Settings/AppSettings.cs:
```diff
@@ -894,6 +894,12 @@ public static bool ShowStashCount
 
+        public static bool ShowAheadBehindData
+        {
+            get => GetBool("showAheadBehindData", true);
```
most keys are lower case only

RussKie replied:
We have a good deal of PascalCased settings as well, but this would be the first camelCased one. 
Agree with @gerhardol, let's align to either be all in a lower case or in a PascalCase

gerhardol commented on GitUI/CommandsDialogs/SettingsDialog/Pages/GeneralSettingsPage.Designer.cs:
```diff
@@ -114,19 +115,20 @@ private void InitializeComponent()
             this.tlpnlPerformance.ColumnStyles.Add(new System.Windows.Forms.ColumnStyle());
             this.tlpnlPerformance.ColumnStyles.Add(new System.Windows.Forms.ColumnStyle());
-            this.tlpnlPerformance.Controls.Add(this.chkCheckForUncommittedChangesInCheckoutBranch, 0, 6);
+            this.tlpnlPerformance.Controls.Add(this.chkShowAheadBehindDataInBrowseWindow, 0, 6);
```
Seem like a bad merge (one option was removed, manual adjustment required)

RussKie commented on UnitTests/GitUITests/UserControls/ToolStripPushButtonTests.cs:
```diff
@@ -0,0 +1,151 @@
+        [OneTimeTearDown]
+        public void OneTimeTearDown()
+        {
+            AppSettings.ShowAheadBehindData = _originalShowAheadBehindData;
```
Just realised, there are no tests that set `AppSettings.ShowAheadBehindData` false

gerhardol commented on GitUI/CommandsDialogs/ToolStripPushButton.cs:
```diff
@@ -0,0 +1,104 @@
+
+            var data = aheadBehindData[branchName];
+
+            if (AppSettings.ShowAheadBehindData)
```
The ahead/behind info is calculated with a separate Git command also if false.
Discussion: If calculate for all local branches, there can be many Git commands. Should this be a separate setting?

gerhardol commented on GitUI/CommandsDialogs/ToolStripPushButton.cs:
```diff
@@ -0,0 +1,104 @@
+
+            ResetToDefaultState();
+
+            var aheadBehindData = _aheadBehindDataProvider.GetData(branchName);
```
$ git for-each-ref --format="%(push:track) %(refname:short)" refs/heads/pmiossec/_display_ahead_behind_push_data 
 pmiossec/_display_ahead_behind_push_data                                                                        

This do obviously not match the regexp and aheadBehindData will be null.

However 
```
$ git for-each-ref --format="%(upstream:track)" refs/heads/pmiossec/_display_ahead_behind_push_data  
[ahead 1]                                                                                            
                                                                                                     
ejgo@ejgo3 MINGW64 /f/dev/gc/gitextensions_4 (pmiossec/_display_ahead_behind_push_data)              
$ git for-each-ref --format="%(push:track)" refs/heads/pmiossec/_display_ahead_behind_push_data      

$
```

gerhardol replied:
upstream works for me, but I do not have push specifically set. The proper solution is to get both upstream and push to get the "best" ahead/behind

gerhardol commented:
Another problem:
For the left panel, only the first level of branches is shown as the wild card is "*" only
```
private IDictionary<string, AheadBehindData> GetData(Encoding encoding, string branchName = "**")
```

gerhardol commented:
Pushed changes that works mostly for me.
* [gone] status is not handled (probably a followup PR)
* If deactivating the service, the pushbutton is not updated. (maybe OK bug)
* The handling of upstream vs push needs some consideration.

RussKie commented on GitCommands/Git/AheadBehindDataProvider.cs:
```diff
@@ -16,22 +16,23 @@ public class AheadBehindDataProvider : IAheadBehindDataProvider
+        // TODO handle [gone] status to show that remote branch no longer exists
         private readonly Regex _aheadBehindRegEx =
-            new Regex(@"^\[(ahead (?<ahead>\d+))?(, )?(behind (?<behind>\d+))?\] (?<branch>.*)$",
+            new Regex(@"^(\[(ahead (?<ahead_u>\d+))?(, )?(behind (?<behind_u>\d+))?\])?::(\[(ahead (?<ahead_p>\d+))?(, )?(behind (?<behind_p>\d+))?\])?::(?<branch>.*)$",
```
Can we please have tests where we get to exercise the full regex and each the left and the right groups?

gerhardol commented:
One optimisation to do, maybe later:
All branchnames are read at least two times now:

Current branch aheadbehind:
for-each-ref --format="%(upstream:track)::%(push:track)::%(refname:short)" refs/heads/pmiossec/_display_ahead_behind_push_data

aheadbehind for side panel (before listing branches):
for-each-ref --format="%(upstream:track)::%(push:track)::%(refname:short)" refs/heads/**

Normal listing of branches and tags for dropdown and side panel (even done twice?):
for-each-ref --sort=-committerdate --sort=-taggerdate --format="%(refname)" refs/

RussKie commented on UnitTests/GitUITests/UserControls/ToolStripPushButtonTests.cs:
```diff
@@ -44,6 +44,8 @@ public void Setup()
         [Test]
         public void DisplayAheadBehindInformation_should_not_display_anything_if_does_not_support_ahead_behind()
         {
+            AppSettings.ShowAheadBehindData = true;
```
We shouldn't be doing this as it modifies a global shared state, and introduces a risk of breaking some other tests that may depend on this.

RussKie commented:
Sent in a tweak and a test fix

gerhardol commented:
Got a NRE, null check required in DisplayAheadBehindInformationForBranches()
Seem to be the same problem that causes tests to fail (will look at the for-each-ref output tomorrow).

at GitUI.BranchTreePanel.RepoObjectsTree.UpdateAheadBehindDataOnbranches(TreeNode treeNode, IDictionary`2 aheadBehindData) in F:\dev\gc\gitextensions_4\GitUI\BranchTreePanel\RepoObjectsTree.cs:line 456
   at GitUI.BranchTreePanel.RepoObjectsTree.DisplayAheadBehindInformationForBranches() in F:\dev\gc\gitextensions_4\GitUI\BranchTreePanel\RepoObjectsTree.cs:line 447
   at GitUI.BranchTreePanel.RepoObjectsTree.<ReloadAsync>d__36.MoveNext() in F:\dev\gc\gitextensions_4\GitUI\BranchTreePanel\RepoObjectsTree.cs:line 174
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable.ConfiguredTaskAwaiter.GetResult()
   at GitUI.ThreadHelper.<>c__DisplayClass13_0.<<FileAndForget>b__0>d.MoveNext() in F:\dev\gc\gitextensions_4\GitExtUtils\GitUI\ThreadHelper.cs:line 101

gerhardol commented:
> NRE

The problem occurs when there are no branches tracked at all

```
$ git for-each-ref --format="%(upstream:track)::%(push:track)::%(refname:short)" refs/heads/
::::master
````

gerhardol commented:
I pushed a correction for the NRE that also simplifies a little and adds a few comments.
The additional for-each-ref should be removed in a follow up issue. Left panel see some changes in #5669 anyway.

RussKie commented:
The tests are still failing... 😞

gerhardol commented:
> The tests are still failing

Because I intentionally just pushed them to my repo, letting @pmiossec decide if they are good.
(I have setup AppVeyor, it builds.)
Commits should be squashed anyway.

gerhardol commented:
Then I will push my changes to your branch.When @RussKie is happy I can squash to two commits or open a new pr. 

Nice job

RussKie commented:
@gerhardol please merge if you're happy with it.

gerhardol commented:
I looked a little reducing the number of for-each-ref (2 calls added to the current 1), but that will not fit in to the left-panel handling. Maybe later.

Thanks @pmiossec

pmiossec commented:
>Thanks @pmiossec

Also, Thanks to you.

RussKie commented:
Thank you all. It is a good feature to have.

francoas commented:
I try to be more specific.

I would add the "not integrated commit" near the **PULL** button, not the push button
and if there are not integrated commits I would change the color of the PULL button, 
because THIS IS THE ACTION (push) that you have to perform to align things

So, in my idea:
 "not integrated commit" - **PULL** button, change PULL button color if != 0, no need fot the arrow
"not pushed commits" - **PUSH** button, PUSH button color if != 0, no need for the arrow
 "things to commit" - **COMMIT** button (as it already does), COMMIT color (as it already does), no need for the word "_commit_"

Samples:

![image](https://user-images.githubusercontent.com/3007941/51306672-39b57a80-1a3e-11e9-8b9f-262b5f5a7bbd.png)


![image](https://user-images.githubusercontent.com/3007941/51306627-1ab6e880-1a3e-11e9-87b2-cea513793364.png)



Forgive me if sometimes I try to see things from the user's point of view and not from the developer's one.

RussKie commented:
> Forgive me if sometimes I try to see things from the user's point of view
and not from the developer's one

That's how it should be, we're building an app for our users

On Thu, Jan 17, 2019, 7:58 PM Franco Andrea <notifications@github.com wrote:

> I try to be more specific.
>
> I would add the "not integrated commit" near the *PULL* button, not the
> push button
> and if there are not integrated commits I would change the color of the
> PULL button,
> because THIS IS THE ACTION (push) that you have to perform to align things
>
> So, in my idea:
> "not integrated commit" - *PULL* button, change PULL button color if !=
> 0, no need fot the arrow
> "not pushed commits" - *PUSH* button, PUSH button color if != 0, no need
> for the arrow
> "things to commit" - *COMMIT* button (as it already does), COMMIT color
> (as it already does), no need for the word "*commit*"
>
> Samples:
>
> [image: image]
> <https://user-images.githubusercontent.com/3007941/51306672-39b57a80-1a3e-11e9-8b9f-262b5f5a7bbd.png>
>
> [image: image]
> <https://user-images.githubusercontent.com/3007941/51306627-1ab6e880-1a3e-11e9-87b2-cea513793364.png>
>
> Forgive me if sometimes I try to see things from the user's point of view
> and not from the developer's one.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/5994#issuecomment-455093090>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEMyXv-otxWM37_AGY4mPJurfUBC_PNNks5vEDsXgaJpZM4Zh-86>
> .
>
